llama-index-core/llama_index/core/workflow/workflow.py

```diff
@@ -0,0 +1,316 @@
+import asyncio
+import warnings
+from collections import defaultdict
+from typing import Any, Callable, Dict, List, Optional, Set
+
+from llama_index.core.workflow.decorators import step, StepConfig
+from llama_index.core.workflow.events import StartEvent, StopEvent, Event
+from llama_index.core.workflow.utils import get_steps_from_class
+
+
+from .errors import WorkflowRuntimeError, WorkflowTimeoutError, WorkflowValidationError
+
+
+class Workflow:
+    def __init__(
+        self,
+        timeout: int = 10,
+        disable_validation: bool = False,
+        verbose: bool = False,
+    ) -> None:
+        # configuration
+        self._timeout = timeout
+        self._disable_validation = disable_validation
+        self._step_flags: Dict[str, asyncio.Event] = {}
+        self._verbose = verbose
+
+        # state variables
+        self._queues: Dict[str, asyncio.Queue] = {}
+        self._tasks: Set[asyncio.Task] = set()
+        self._events: List[Any] = []
+        self._event_log: List[Event] = []
+        self._retval = None
+        self._event_subscriptions: Dict[type, Set[str]] = defaultdict(set)
+
+    def _start(self, stepwise: bool = False) -> None:
+        """Sets up the queues and tasks for each declared step.
+
+        This method also launches each step as an async task.
+        """
+        for name, step_func in get_steps_from_class(self).items():
+            self._queues[name] = asyncio.Queue()
+            self._step_flags[name] = asyncio.Event()
+            step_config: Optional[StepConfig] = getattr(
+                step_func, "__step_config", None
+            )
+            if not step_config:
+                raise ValueError(f"Step {name} is missing `@step()` decorator.")
+
+            async def _task(
+                name: str,
+                queue: asyncio.Queue,
+                step: Callable,
+                config: StepConfig,
+            ) -> None:
+                while True:
+                    ev = await queue.get()
+                    if type(ev) not in config.accepted_events:
+                        continue
+
+                    # do we need to wait for the step flag?
+                    if stepwise:
+                        await self._step_flags[name].wait()
+
+                        # clear all flags so that we only run one step
+                        for flag in self._step_flags.values():
+                            flag.clear()
+
+                    if self._verbose:
+                        print(f"Running step {name}")
+
+                    # run step
+                    new_ev = await step(ev)
+
+                    if self._verbose:
+                        print(f"Step {name} produced event {new_ev}")
+
+                    # handle the return value
+                    if new_ev is None:
+                        continue
+
+                    # Store the accepted event for the drawing operations
+                    self._events.append((name, type(ev).__name__))
```

Comment:
I think this implies that steps can only return a single event right?

Reply:
I was about to add a note in the PR body - while technically possible, I wanted to discuss with you which use cases might benefit of sending back multiple events, which incidentally is still possible by calling `self.send_event` from a step function